rir/src/compiler/pir/instruction.h

```diff
@@ -412,25 +412,42 @@ class Instruction : public Value {
     PirType inferredTypeForArithmeticInstruction(const GetType& getType) const {
         auto m = mergedInputType(getType);
         if (!m.maybeObj()) {
-            auto t = PirType::bottom();
+            // Everything but numbers throws an error
+            auto t = m & PirType::num().notMissing().orAttribsOrObj();
+
+            // TRUE + TRUE == 2
+            if (m.maybe(RType::logical))
+                t = t.orT(RType::integer).notT(RType::logical);
+
+            // If either argument is complex the result will be complex
+            // If one or both arguments are numeric, the result will
+            // be numeric
             eachArg([&](Value* v) {
-                if (!mayHaveEnv() || v != env())
-                    t = t.mergeWithConversion(getType(v));
+                if (getType(v).isA(RType::cplx))
+                    t = t.notT(RType::real).notT(RType::integer);
+                if (getType(v).isA(RType::real))
+                    t = t.notT(RType::integer);
             });
-            // Everything but numbers throws an error
-            t = t & PirType::num().notMissing().orAttribsOrObj();
-            // e.g. TRUE + TRUE == 2
-            if (m.maybe(RType::logical)) {
-                t = t.orT(RType::integer);
-                t = t.notT(RType::logical);
-            }
-            // the binop result becomes NA if it can't be represented in a
+
+            // The binop result becomes NA if it can't be represented in a
             // fixpoint integer (e.g. INT_MAX + 1 == NA)
-            // * the condition checks iff at least one of the arguments is an
+            // The condition checks if at least one of the arguments is an
             // integer (doesn't happen with only logicals), and the result is an
             // integer (doesn't happen with real coercion)
             if (m.maybe(RType::integer) && t.maybe(RType::integer))
```

Comment:
uhm, could this interact with the change of int->real below?

Reply:
Good catch, thanks!